src/byte_set.rs

```diff
@@ -126,21 +133,55 @@ impl<const LEN: usize> From<[u8; LEN]> for ByteSet {
     }
 }
 
-impl From<Range<u8>> for ByteSet {
-    fn from(range: Range<u8>) -> ByteSet {
+impl<'a> From<&'a [u8]> for ByteSet {
+    fn from(bytes: &'a [u8]) -> ByteSet {
         let mut bs = ByteSet::new();
-        for b in range {
-            bs.insert(b);
+        for b in bytes {
+            bs.insert(*b);
         }
         bs
     }
 }
 
-impl From<RangeInclusive<u8>> for ByteSet {
-    fn from(range: RangeInclusive<u8>) -> ByteSet {
-        let mut bs = ByteSet::new();
-        for b in range {
-            bs.insert(b);
+impl From<ops::Range<u8>> for ByteSet {
+    fn from(value: ops::Range<u8>) -> Self {
+        ByteSet::from(value.start..=value.end - 1)
+    }
+}
+
+impl From<ops::RangeInclusive<u8>> for ByteSet {
+    fn from(value: ops::RangeInclusive<u8>) -> Self {
+        // because the values are adjacent, we can optimize if they are within the same quadrant
+        let lo = value.start();
+        let hi = value.end();
+        let start_ix = *lo / 64;
+        let end_ix = *hi / 64;
+        if start_ix == end_ix {
+            let mut mask = 0u64;
+            for i in value {
+                mask |= 1 << i % 64;
+            }
+            let bits = match start_ix {
+                0 => [mask, 0, 0, 0],
+                1 => [0, mask, 0, 0],
+                2 => [0, 0, mask, 0],
+                3 => [0, 0, 0, mask],
+                _ => unreachable!("u8 / 64 not in range 0..=3 ??"),
+            };
+            ByteSet { bits }
+        } else {
+            // Fall-back to FromIter implementation, which is less efficient in general
+            value.collect::<ByteSet>()
+        }
+    }
+}
```

Comment:
I think this is overkill; ranged byteset construction is not in the fast path. :sweat_smile:

Reply:
I use it a lot for UTF-8.